scrapers/fl/utils.py

```diff
@@ -1,6 +1,142 @@
+import logging
+import random
+import time
+from http.client import RemoteDisconnected
+from urllib.error import URLError
+
+from requests.exceptions import ConnectionError, Timeout, RequestException
+
+logger = logging.getLogger(__name__)
+
+
 def fix_name(name):
     # handles cases like Watson, Jr., Clovis
     if ", " not in name:
         return name
     last, first = name.rsplit(", ", 1)
     return first + " " + last
+
+
+def get_random_user_agent():
+    """
+    Return a random user agent to help avoid detection.
+    """
+    user_agents = [
+        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36",
+        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.1.1 Safari/605.1.15",
+        "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:89.0) Gecko/20100101 Firefox/89.0",
+        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.107 Safari/537.36",
+        "Mozilla/5.0 (iPhone; CPU iPhone OS 14_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1",
+        "Mozilla/5.0 (iPad; CPU OS 14_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1",
+        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36 Edg/91.0.864.59",
+    ]
+    return random.choice(user_agents)
+
+
+def add_random_delay(min_seconds=1, max_seconds=3):
+    """
+    Add a random delay to simulate human behavior.
+
+    Args:
+        min_seconds: Minimum delay in seconds
+        max_seconds: Maximum delay in seconds
+    """
+    delay = random.uniform(min_seconds, max_seconds)
+    logger.debug(f"Adding random delay of {delay:.2f} seconds")
+    time.sleep(delay)
+
+
+def retry_on_connection_error(func, max_retries=5, initial_backoff=2, max_backoff=60):
```

Comment:
(also in broader context re: scrapy: seems like this logic could fit neatly between scrapy retry settings https://docs.scrapy.org/en/latest/topics/downloader-middleware.html#std-setting-RETRY_TIMES and a custom downloader middleware https://github.com/scrapy/scrapy/blob/master/docs/topics/downloader-middleware.rst)